safehttp/mux_test.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/safehtml"
+	"github.com/google/safehtml/template"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+)
+
+type fakeDispatcher struct{}
+
+func (fakeDispatcher) Write(rw http.ResponseWriter, resp Response) error {
+	switch x := resp.(type) {
+	case safehtml.HTML:
+		_, err := rw.Write([]byte(x.String()))
+		return err
+	default:
+		panic("not a safe response type")
+	}
+}
+
+func (fakeDispatcher) ExecuteTemplate(rw http.ResponseWriter, t Template, data interface{}) error {
+	switch x := t.(type) {
+	case *template.Template:
+		return x.Execute(rw, data)
+	default:
+		panic("not a safe response type")
+	}
+}
+
+type fakeResponseWriter struct {
+	header http.Header
+	writer io.Writer
+	status int
+}
+
+func newFakeResponseWriter(w io.Writer) *fakeResponseWriter {
+	return &fakeResponseWriter{header: http.Header{}, writer: w, status: http.StatusOK}
+}
+
+func (r *fakeResponseWriter) Header() http.Header {
+	return r.header
+}
+
+func (r *fakeResponseWriter) WriteHeader(statusCode int) {
+	r.status = statusCode
+}
+
+func (r *fakeResponseWriter) Write(data []byte) (int, error) {
+	return r.writer.Write(data)
+}
+
+func TestValidHost(t *testing.T) {
```

Comment:
I think that this and the following one-handler tests might benefit from having a table test that has

```go
tests := []struct{
  request  *http.Request
  wantStatus int
  wantHeader map[string][]string
  wantBody string
} {...}
```
Otherwise it is very hard to see the difference between the various tests and what they are there for.